File: src/LMVideoStudio.Client/src-js/ProjectJson.js
import { SequencePresetModule_toSchemaValue, Project, TransitionSpecModule_defaultMockup, RenderDefaults, RenderProfileModule_defaultBake, RenderProfileModule_defaultMockup, ProjectModule_defaultMockupDurationSec, SequencePreset, SequencePresetModule_fromSchemaValue, StoryboardBlock, BlockGeneration, BlockSource, TransitionType } from "./LMVideoStudio.Domain/Types.js";
import { fromString, float, list as list_2, map, string, int, guid, object } from "./fable_modules/Thoth.Json.10.4.1/Decode.fs.js";
import { equals, uncurry2 } from "./fable_modules/fable-library-js.4.27.0/Util.js";
import { map as map_1, filter, defaultArg } from "./fable_modules/fable-library-js.4.27.0/Option.js";
import { map as map_2, toArray, singleton as singleton_1, isEmpty, empty } from "./fable_modules/fable-library-js.4.27.0/List.js";
import { toString, guid as guid_1, object as object_1 } from "./fable_modules/Thoth.Json.10.4.1/Encode.fs.js";
import { singleton, append, delay, toList } from "./fable_modules/fable-library-js.4.27.0/Seq.js";

function transitionTypeCodec(_arg) {
    switch (_arg) {
        case "cut":
            return new TransitionType(0, []);
        case "fade":
            return new TransitionType(1, []);
        case "crossfade":
            return new TransitionType(2, []);
        case "slide_left":
            return new TransitionType(3, []);
        case "slide_right":
            return new TransitionType(4, []);
        case "dissolve":
            return new TransitionType(5, []);
        default:
            throw new Error(`Unknown transition: ${_arg}`);
    }
}

function blockSourceCodec(_arg) {
    switch (_arg) {
        case "imported":
            return new BlockSource(0, []);
        case "generated":
            return new BlockSource(1, []);
        default:
            throw new Error(`Unknown source: ${_arg}`);
    }
}

const blockDecoder = (path_16) => ((v_1) => object((get$) => {
    let objectArg, objectArg_1, objectArg_2, objectArg_3, objectArg_4, objectArg_5, objectArg_6, objectArg_7, objectArg_8, objectArg_9, objectArg_10, objectArg_11;
    return new StoryboardBlock((objectArg = get$.Required, objectArg.Field("id", guid)), (objectArg_1 = get$.Required, objectArg_1.Field("order", uncurry2(int))), (objectArg_2 = get$.Optional, objectArg_2.Field("title", string)), (objectArg_3 = get$.Required, objectArg_3.Field("source", (path_3, value_3) => map(blockSourceCodec, string, path_3, value_3))), (objectArg_4 = get$.Optional, objectArg_4.Field("thumbnailPath", string)), (objectArg_5 = get$.Optional, objectArg_5.Field("imagePrompt", string)), (objectArg_6 = get$.Optional, objectArg_6.Field("voiceoverScript", string)), (objectArg_7 = get$.Optional, objectArg_7.Field("directorNotes", string)), defaultArg((objectArg_8 = get$.Optional, objectArg_8.Field("moodTags", (path_9, value_9) => list_2(string, path_9, value_9))), empty()), (objectArg_9 = get$.Optional, objectArg_9.Field("mockupDurationSec", float)), (objectArg_10 = get$.Optional, objectArg_10.Field("bakeDurationSec", float)), undefined, undefined, (objectArg_11 = get$.Optional, objectArg_11.Field("generation", (path_15, v) => object((g) => {
        let objectArg_12, objectArg_13, objectArg_14;
        return new BlockGeneration((objectArg_12 = g.Optional, objectArg_12.Field("seed", uncurry2(int))), (objectArg_13 = g.Optional, objectArg_13.Field("referenceAssetPath", string)), filter((arg_30) => !isEmpty(arg_30), (objectArg_14 = g.Optional, objectArg_14.Field("thumbnailVariants", (path_14, value_15) => list_2(string, path_14, value_15)))));
    }, path_15, v))), undefined);
}, path_16, v_1));

export const projectDecoder = (path_6) => ((v) => object((get$) => {
    let objectArg, objectArg_1, objectArg_2, objectArg_3, objectArg_4, objectArg_5, objectArg_6;
    const preset = defaultArg(SequencePresetModule_fromSchemaValue((objectArg = get$.Required, objectArg.Field("sequencePreset", string))), new SequencePreset(0, []));
    return new Project((objectArg_1 = get$.Required, objectArg_1.Field("schemaVersion", uncurry2(int))), (objectArg_2 = get$.Required, objectArg_2.Field("id", guid)), (objectArg_3 = get$.Required, objectArg_3.Field("name", string)), undefined, undefined, (objectArg_4 = get$.Optional, objectArg_4.Field("brief", string)), preset, defaultArg((objectArg_5 = get$.Optional, objectArg_5.Field("defaultMockupDurationSec", float)), ProjectModule_defaultMockupDurationSec), new RenderDefaults(RenderProfileModule_defaultMockup, RenderProfileModule_defaultBake), undefined, defaultArg((objectArg_6 = get$.Optional, objectArg_6.Field("blocks", (path_5, value_6) => list_2(uncurry2(blockDecoder), path_5, value_6))), empty()), TransitionSpecModule_defaultMockup);
}, path_6, v));

export function blockEncoder(block) {
    return object_1(toList(delay(() => append(singleton(["id", guid_1(block.Id)]), delay(() => append(singleton(["order", block.Order]), delay(() => append(singleton(["source", equals(block.Source, new BlockSource(0, [])) ? "imported" : "generated"]), delay(() => append(defaultArg(map_1((t) => singleton_1(["title", t]), block.Title), empty()), delay(() => defaultArg(map_1((p) => singleton_1(["thumbnailPath", p]), block.ThumbnailPath), empty()))))))))))));
}

export function encodeProject(project) {
    return toString(2, object_1([["schemaVersion", project.SchemaVersion], ["id", guid_1(project.Id)], ["name", project.Name], ["sequencePreset", SequencePresetModule_toSchemaValue(project.SequencePreset)], ["defaultMockupDurationSec", project.DefaultMockupDurationSec], ["renderDefaults", object_1([["mockup", object_1([["tier", "mockup"], ["width", project.RenderDefaults.Mockup.Width], ["height", project.RenderDefaults.Mockup.Height]])], ["bake", object_1([["tier", "bake"], ["width", project.RenderDefaults.Bake.Width], ["height", project.RenderDefaults.Bake.Height]])]])], ["blocks", toArray(map_2(blockEncoder, project.Blocks))]]));
}

export function decodeProject(json) {
    return fromString(uncurry2(projectDecoder), json);
}

File: src/LMVideoStudio.Client/ProjectJson.fs
module LMVideoStudio.Client.ProjectJson

open Thoth.Json
open LMVideoStudio.Domain

[<AutoOpen>]
module Json =
    let private transitionTypeCodec =
        function
        | "cut" -> Cut
        | "fade" -> Fade
        | "crossfade" -> Crossfade
        | "slide_left" -> SlideLeft
        | "slide_right" -> SlideRight
        | "dissolve" -> Dissolve
        | other -> failwith $"Unknown transition: {other}"

    let private blockSourceCodec =
        function
        | "imported" -> BlockSource.Imported
        | "generated" -> BlockSource.Generated
        | other -> failwith $"Unknown source: {other}"

    let private dateTimeOffsetDecoder =
        Decode.string
        |> Decode.andThen (fun s ->
            match System.DateTimeOffset.TryParse s with
            | true, d -> Decode.succeed d
            | _ -> Decode.fail $"Invalid date: {s}")

    let private blockDecoder =
        Decode.object (fun get ->
            { Id = get.Required.Field "id" Decode.guid
              Order = get.Required.Field "order" Decode.int
              Title = get.Optional.Field "title" Decode.string
              Source = get.Required.Field "source" (Decode.string |> Decode.map blockSourceCodec)
              ThumbnailPath = get.Optional.Field "thumbnailPath" Decode.string
              ImagePrompt = get.Optional.Field "imagePrompt" Decode.string
              VoiceoverScript = get.Optional.Field "voiceoverScript" Decode.string
              DirectorNotes = get.Optional.Field "directorNotes" Decode.string
              MoodTags = get.Optional.Field "moodTags" (Decode.list Decode.string) |> Option.defaultValue []
              MockupDurationSec = get.Optional.Field "mockupDurationSec" Decode.float
              BakeDurationSec = get.Optional.Field "bakeDurationSec" Decode.float
              Transitions = None
              Audio = None
              Generation =
                  get.Optional.Field "generation" (Decode.object (fun g ->
                      { Seed = g.Optional.Field "seed" Decode.int
                        ReferenceAssetPath = g.Optional.Field "referenceAssetPath" Decode.string
                        ThumbnailVariants =
                            g.Optional.Field "thumbnailVariants" (Decode.list Decode.string)
                            |> Option.filter (not << List.isEmpty) }))
              Artifacts = None })

    let projectDecoder =
        Decode.object (fun get ->
            let preset =
                get.Required.Field "sequencePreset" Decode.string
                |> SequencePreset.fromSchemaValue
                |> Option.defaultValue Preset1080p24

            { SchemaVersion = get.Required.Field "schemaVersion" Decode.int
              Id = get.Required.Field "id" Decode.guid
              Name = get.Required.Field "name" Decode.string
              CreatedAt = get.Optional.Field "createdAt" dateTimeOffsetDecoder
              UpdatedAt = get.Optional.Field "updatedAt" dateTimeOffsetDecoder
              Brief = get.Optional.Field "brief" Decode.string
              SequencePreset = preset
              DefaultMockupDurationSec =
                  get.Optional.Field "defaultMockupDurationSec" Decode.float
                  |> Option.defaultValue Project.defaultMockupDurationSec
              RenderDefaults =
                  { Mockup = RenderProfile.defaultMockup
                    Bake = RenderProfile.defaultBake }
              StylePack = None
              Blocks = get.Optional.Field "blocks" (Decode.list blockDecoder) |> Option.defaultValue []
              TransitionsDefault = Some TransitionSpec.defaultMockup })

    let blockEncoder (block: StoryboardBlock) =
        Encode.object [
            "id", Encode.guid block.Id
            "order", Encode.int block.Order
            "source", Encode.string (if block.Source = BlockSource.Imported then "imported" else "generated")
            yield!
                (block.Title
                 |> Option.map (fun t -> [ "title", Encode.string t ])
                 |> Option.defaultValue [])
            yield!
                (block.ThumbnailPath
                 |> Option.map (fun p -> [ "thumbnailPath", Encode.string p ])
                 |> Option.defaultValue [])
        ]

    let encodeProject (project: Project) =
        Encode.object [
            "schemaVersion", Encode.int project.SchemaVersion
            "id", Encode.guid project.Id
            "name", Encode.string project.Name
            "sequencePreset", Encode.string (SequencePreset.toSchemaValue project.SequencePreset)
            "defaultMockupDurationSec", Encode.float project.DefaultMockupDurationSec
            "renderDefaults",
            Encode.object [
                "mockup",
                Encode.object [
                    "tier", Encode.string "mockup"
                    "width", Encode.int project.RenderDefaults.Mockup.Width
                    "height", Encode.int project.RenderDefaults.Mockup.Height
                ]
                "bake",
                Encode.object [
                    "tier", Encode.string "bake"
                    "width", Encode.int project.RenderDefaults.Bake.Width
                    "height", Encode.int project.RenderDefaults.Bake.Height
                ]
            ]
            "blocks", project.Blocks |> List.map blockEncoder |> List.toArray |> Encode.array
        ]
        |> Encode.toString 2

    let decodeProject json = Decode.fromString projectDecoder json
